src/router/router.js

```diff
@@ -0,0 +1,104 @@
+class Router {
+  #routes = [];
+  #rootElement;
+  #component;
+  #params;
+
+  constructor(id) {
+    this.#rootElement = document.getElementById(id);
+
+    window.addEventListener('hashchange', this.initComponent, false);
+
+    return this;
+  }
+
+  get params() {
+    return this.#params.slice(1);
+  }
+
+  init() {
+    this.initComponent();
+
+    return this;
+  }
+
+  add(path, component) {
+    const pathRegExp = new RegExp('^' + path.replace(/:[^\s/]+/g, '([\\w-]+)') + '$');
+
+    this.#routes.push({
+      pathRegExp,
+      component,
+    });
+
+    return this;
+  }
+
+  push(path) {
+    const route = this._getRouteByPath(path);
+
+    if (route) {
+      location.hash = path;
+    }
+
+    return this;
+  }
+
+  initComponent = () => {
+    const path = location.hash.slice(1);
+    const route = this._getRouteByPath(path);
+
+    if (!route) {
+      return;
+    }
+
+    if (this.#component) {
+      this.componentWillUnmount();
+    }
+
+    this.#params = this._getParamsByPath(route.pathRegExp, path);
+    this.#component = new route.component({ apply: this.componentRender, router: this });

```

Comment:
`apply: this.componentRender` I hope/believe it will be removed after changes in router

Reply:
done